powershell completion with custom comp

Luap99 opened the pull request:
The current powershell completion is not very capable.

Let's port it to the go custom completion logic to have a
unified experience accross all shells.

Powershell supports three different completion modes

- TabCompleteNext (default windows style - on each key press the next option is displayed)
- Complete (works like bash)
- MenuComplete (works like zsh)

You set the mode with `Set-PSReadLineKeyHandler -Key Tab -Function <mode>`

Descriptions will only be supported for Complete and MenuComplete.

Fixes #1229

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
 
 // GenPowerShellCompletion generates PowerShell completion file and writes to the passed writer.
-func (c *Command) GenPowerShellCompletion(w io.Writer) error {
+func (c *Command) GenPowerShellCompletion(w io.Writer, includeDesc bool) error {
```
This change is not backwards compatible.  We'll need to add new functions instead.

Luap99 replied:
Yes I'm aware. I'm waiting for #1146 to decide how we want to handle this.

marckhouzam replied:
Since we know this change cannot be done, I suggest you implement whatever API you prefer and restore the previous methods.  This will at least put the PR in a state that can be merged.

Luap99 commented:
@marckhouzam I don't think I can get `ShellCompDirectiveFilterFileExt` and `ShellCompDirectiveFilterDirs` to work, I guess the best way is to provide the standard path completion instead?

marckhouzam commented:
@Luap99 agreed. That's what I also had to do for fish completion:
https://github.com/spf13/cobra/blob/02a0d2fbc9e61d26f8e5979749f6030964a55a3e/fish_completions.go#L126-L133

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+        # Set the description to a one space string if there is none set.
+        # This is needed because the CompletionResult does not accept an empty string as argument
+        if (-Not $Description) {
+            $Description = " "
```
I notice that if there are no descriptions, we see a set of empty parentheses.  It is not so bad, but fish does not put them, but from the comment, I guess you don't have a choice?  For example:

powershell:
```
> ./testprog no<TAB>
nodesc  ( )                        nofile  (No file completion done)
```
fish:
```
> ./testprog no<TAB>
nodesc  nofile  (No file completion done)
```

marckhouzam replied:
This is particularly visible when I use a `--no-descriptions` flag:
```
> ./testprog
cmdColonArgs                     ( )  help                             ( )
cmdLongDesc12345678901234567890  ( )  nodesc                           ( )
compAndFile                      ( )  nofile                           ( )
completion                       ( )  specialChar                      ( )
dirOnly                          ( )  subdir                           ( )
filterFile                       ( )  vargs                            ( )
fullFile                         ( )
```

Luap99 replied:
That should be no problem to fix

Luap99 replied:
Ok that's working now.

SilkyFowl replied:
Good!!

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+
+        $Values | ForEach-Object {
+            __%[1]s_debug "Filter: $($_.Name)"
+            Get-ChildItem -Recurse -Name -File -Filter "*.$($_.Name)"
```
This seems to work pretty decently.  I thought you said `ShellCompDirectiveFilterFileExt` wasn't working?

Also, why to you recurse into sub directories?

Luap99 replied:
Yeah it works kinda, but I can't make it handle paths correctly. I can only make it complete the full path or only the file name. It won't work with relative paths. It also breaks completely with spaces.

That's why I'm thinking that sticking to standard path completion done by powershell would be better. This is just a ugly hack and doesn't seam to be supported by powershell in any way.

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    __%[1]s_debug "The completion directive is: $Directive"
+    
+    # remove directive (last element) from out
+    $Out = $Out | Where-Object { $_ –ne $Out[-1] }
```
My VSCode tells me the `-` in `-ne` in the above line is not a real `-`.  Was that on purpose?

Luap99 replied:
No, some copy-paste problem I guess

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    # The user could have moved the cursor backwards on the command-line.
+    # We need to trigger completion from the $CursorPosition location, so we need
+    # to truncate the command-line ($Command) up to the $CursorPosition location.
+    # Make sure the $Command is longer then the $CursorPosition befor we truncate.
```
`befor` missing the final `e`

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    $ShellCompDirectiveFilterDirs=%[7]d
+
+	# Prepare the command to request completions for the program.
+    # Split the command at the first space to seperate the programm and arguments
```
`separate` instead of `seperate`

and `programm` with two `m`

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    $RequestComp="$Program %[2]s $Arguments"
+    __%[1]s_debug "RequestComp: $RequestComp"
+
+    # we cannont use $WordToComplete because it 
```
`cannot` instead of `cannont`

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+
+            # TabCompleteNext and in case we get something unknown
+            Default {
+                # Like MenuComplete but we dont want to add a space here because 
```
`don't` instead of `dont`

(it's my VSCode that's reporting me these typos)

Luap99 replied:
I also use VSCode, is that an extension? It seams like I need that one too.

marckhouzam replied:
I use the extension "Code Spell Checker"

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    __%[1]s_debug "RequestComp: $RequestComp"
+
+    # we cannont use $WordToComplete because it 
+    # has the wrong values if the curser was moved
```
`cursor` instead of `curser`

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+    # get directive from last line
+    [int]$Directive = $Out[-1].TrimStart(':')
+    if ($Directive -eq "") {
+        # There is not directive specified
```
`no` instead of `not` (that was my typo in the bash script 😄 )

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+        # - TabCompleteNext (default windows style - on each key press the next option is displayed)
+        # - Complete (works like bash)
+        # - MenuComplete (works like zsh)
+        # You set the mode with Set-PSReadLineKeyHandler -Key Tab -Function <mode>
```
Super nice!  Each mode works great!

Luap99 commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
-		usage := escapeStringForPowerShell(subCmd.Short)
-		fmt.Fprintf(out, "\n            [CompletionResult]::new('%s', '%s', [CompletionResultType]::ParameterValue, '%s')", subCmd.Name(), subCmd.Name(), usage)
+function __%[1]s_debug {
+	if (Get-Variable "POWERSHELL_COMP_DEBUG_FILE" -Scope Global) {
```
OK, I will change this

Luap99 commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
-		fmt.Fprintf(out, "\n            [CompletionResult]::new('%s', '%s', [CompletionResultType]::ParameterValue, '%s')", subCmd.Name(), subCmd.Name(), usage)
+function __%[1]s_debug {
+	if (Get-Variable "POWERSHELL_COMP_DEBUG_FILE" -Scope Global) {
+		"$args" | Out-File -Append -FilePath "${POWERSHELL_COMP_DEBUG_FILE}"
```
Ah yeah, it's not a env var. You have to set $POWERSHELL_COMP_DEBUG_FILE as normal var in the shell. I will change this to an env var to suit the other shells and in cobra debugging.

Luap99 commented on powershell_completions.go:
```diff
@@ -11,90 +8,253 @@ import (
+                    [System.Management.Automation.CompletionResult]::new("$($comp.Name)$Space", "$($comp.Name)", 'ParameterValue', "$($comp.Description)")
+
+                } else {
+                    ##TODO Trim Description if to long
```
Nice, I didn't even consider this for a sec :sweat_smile:

Luap99 commented:
@marckhouzam Thanks for the fast review. I ignore `ShellCompDirectiveFilterFileExt` and `ShellCompDirectiveFilterDirs` now because there to many corner cases I can't handle.

marckhouzam commented:
@Luap99 Everything works great!  As you said, the `--flag=` form would be nice, but personally, I don't think it should be a blocker. The documentation is important but after that, I would love to see this merged.

marckhouzam commented:
Oh, and I think you should remove `powershell_completions_test.go`.
I'm not sure if there are tests you can implement since there is no more generation of code besides the fixed script.
I removed `zsh_completions_test.go` when I did the zsh port to Go completions.

Luap99 commented:
> Luap99 Everything works great! As you said, the `--flag=` form would be nice, but personally, I don't think it should be a blocker. The documentation is important but after that, I would love to see this merged.

I'm still thinking how this could be done. Powershell does not have `COMP_WORDBREAKS`, so I can only imagine something like this.
```
$ ./testprog --flag=[TAB]
--flag=foo
--flag=bar
--flag=foobar
```
Not very nice but it could work.

marckhouzam commented:
> I'm still thinking how this could be done. Powershell does not have `COMP_WORDBREAKS`, so I can only imagine something like this.
> 
> ```
> $ ./testprog --flag=[TAB]
> --flag=foo
> --flag=bar
> --flag=foobar
> ```
> 
> Not very nice but it could work.

This is what `helm` and `kubectl` currently offer for their zsh completion (not yet migrated to the new Cobra zsh v2):
```
kubectl --context=<TAB>
--context=acc       --context=cor       --context=dev       --context=lab       --context=minikube
--context=bi.dev    --context=default   --context=infra     --context=lab2      --context=prod
```

I think it is just fine.

Luap99 commented:
> This is what helm and kubectl currently offer for their zsh completion (not yet migrated to the new Cobra zsh v2):

I'm curios does that work with file path completion?

I got it working for powershell with arguments, but there is no way to make it work file completion.

Luap99 commented:
Ok, I think the script is in good shape now. I added documentation.

One question remains, how should the API look like?

marckhouzam commented:
> > This is what helm and kubectl currently offer for their zsh completion (not yet migrated to the new Cobra zsh v2):
> 
> I'm curious does that work with file path completion?

I hadn't realized, but no, it does not work with file completion.  So, currently, for zsh (not using Cobra's zsh support):
```
$ kubectl --kubeconfig /User<TAB>     # Works
$ kubectl --kubeconfig=/User<TAB>     # Does not work
```
> I got it working for powershell with arguments, but there is no way to make it work with file completion.

I think that is fine.

marckhouzam commented on shell_completions.md:
```diff
@@ -431,4 +439,38 @@ search  show  status
 ## PowerShell completions
 
-Please refer to [PowerShell Completions](powershell_completions.md) for details.
+Cobra supports native PowerShell completions generated from the root `cobra.Command`. You can use the `command.GenPowerShellCompletion()` or `command.GenPowerShellCompletionFile()` functions. You must provide these functions with a parameter indicating if the completions should be annotated with a description. Cobra will provide the description automatically based on usage information. You can ch ...
```
This needs to be update to follow the API that will be chosen (which cannot be this one).

marckhouzam commented on shell_completions.md:
```diff
@@ -431,4 +439,38 @@ search  show  status
+* TabCompleteNext (default windows style - on each key press the next option is displayed)
+* Complete (works like bash)
+* MenuComplete (works like zsh)
+You set the mode with `Set-PSReadLineKeyHandler -Key Tab -Function <mode>`. Descriptions are only displayed when using the `Complete` or `MenuComplete` mode.
```
Need a newline before this line.  GitHub currently puts on the same line as the last bullet, can be seen here:
https://github.com/Luap99/cobra/blob/powershell-custom-comp/shell_completions.md

marckhouzam commented on powershell_completions.go:
```diff
@@ -11,90 +8,254 @@ import (
 
 // GenPowerShellCompletionFile generates PowerShell completion file.
-func (c *Command) GenPowerShellCompletionFile(filename string) error {
+func (c *Command) GenPowerShellCompletionFile(filename string, includeDesc bool) error {
```
To make sure we don't forget, this must be reverted as it breaks compatibility.

Luap99 commented:
I just tested this on windows 10 with the default powershell v5.1 and I noticed one problem. `testprog.exe --[TAB]` doesn't work but I think this is not a bug in my script since Powershell doesn't call the completion script at all (there is no debug output). Also ```testprog.exe `--[TAB]``` and `testprog.exe --r[TAB]` are working. Since `testprog.exe --[TAB]` works in Powershell v7, I call this a bug in Powershell v5.1.
I think there is nothing I can do other then adding a note in the docs.

Apart from that, I think it works very well.

jpmcb commented:
I have no way of testing this on a windows machine. Are there other's in the community that would be able to lend a hand and test this out?

marckhouzam commented:
> I just tested this on windows 10 with the default powershell v5.1 and I noticed one problem. `testprog.exe --[TAB]` doesn't work but I think this is not a bug in my script since Powershell doesn't call the completion script at all (there is no debug output). Also `` testprog.exe `--[TAB] `` and `testprog.exe --r[TAB]` are working. Since `testprog.exe --[TAB]` works in Powershell v7, I call this a bug in Powershell v5.1.
> I think there is nothing I can do other then adding a note in the docs.

I can confirm I see this same behavior with powershell 5.1.  I think we'll just have to live with it.

I've tested on Mac with `pwsh 7.0.2` and on Windows 10 with `powershell 5.1` by using this PR with `helm`.  
Everything works really well.
Beyond standard completions, I've tested:
* `ShellDirectiveNoSpace`
* `ShellDirectiveNoFileComp`
* flag completion with the `=` form and the normal form
* descriptions
* all three types of completions for powershell

From a functionality perspective, not only do I feel this is complete, but I believe it is dramatically better than what we have right now for powershell.

@jpmcb There is still a need for guidance on what form the API should take for programs to call this.

Luap99 commented:
@marckhouzam Thanks for testing.

I added one small thing to properly escape spacial chars with a backtick. So they can safely be used as input.

I also changed the API to `GenPowerShellCompletion()` and `GenPowerShellCompletionWithDesc()` to keep it backwards compatible.

marckhouzam reviewed (approved):
Very nice.
Could you just add `powershell` in `shell_completions.md` in the `Descriptions for completions` section?

rshoemaker commented:
Hi,

I've been experimenting with your patch and think I've found an issue with it.  If I understand how your patch works, it seems to be generating the completions by interrogating the binary itself via "--help" or usage - I'm not really sure.  The issue with that is that you might have a different command tree that you want to generate completions for.  In my case, we have a complicated Cobra command tree that spans multiple Go binaries.  One binary serves as the core CLI and then we have add-on modules that provide additional sub-commands.  In order to generate completion bindings for Bash and ZSH, we generate a superset of the Cobra CLI commands and then call the GenXXXCompletion() funcs to create a completion file that spans all of our binaries (but they all share a common Cobra root command).  Your patch uses the CLI visible only in the current binary, which may not be the command you called GenXXXCompletion() func on. 

The attached example illustrates the issue.  We have a "root-cmd" with one sub-command "child-cmd".  Then we have an additional Cobra command tree that contains "root-cmd", "child-cmd", and "generated-cmd" (representing a command from another binary).  When we generate Powershell completions on the additional Cobra command tree using your patch, we only get completions for "child-cmd" and NOT "generated-cmd".  This patch seems to break the semantics of the GenXXXCompletion() func and is different from the approach used in the ZSH and Bash versions of the func.

[main.zip.zip](https://github.com/spf13/cobra/files/5271247/main.zip.zip)

rshoemaker commented:
And by the way, we couldn't get powershell completion working at all without your patch, so it's a big improvement.  We'd like to see this get merged back in, especially if you can tweak it to generate completions for the specified command tree.

Luap99 commented:
@rshoemaker Thanks for your feedback. Your specific problem is not directly related to this PR. To see how the completions are provided see: https://github.com/spf13/cobra/blob/master/shell_completions.md#debugging. This script only parses this output and provides the completions. Currently (master branch) zsh and fish are already using this type of completion script and will face the same problem. This PR only adds support for powershell.

However @marckhouzam (the original author for this type of completions) has a PR which I believe could add support for your use case https://github.com/spf13/cobra/pull/1161#issuecomment-674427951. He can properly clarify.

Luap99 commented:
Also @marckhouzam do you think it would make sense to disable `GenXXXCompletion()` on child commands for custom go completion scripts? This only results in a non working completion script.

rshoemaker commented:
HI @Luap99 , @marckhouzam ,

Thanks for the quick reply.  I ran some more experiments where I dump all the shell completions for our root command and our generated command tree:

`	_ = RootCmd.GenPowerShellCompletionFile("root.pwr")
	_ = GeneratedCli.GenPowerShellCompletionFile("gen.pwr")

	_ = RootCmd.GenZshCompletionFile("root.zsh")
	_ = GeneratedCli.GenZshCompletionFile("gen.zsh")

	_ = RootCmd.GenBashCompletionFile("root.bsh")
	_ = GeneratedCli.GenBashCompletionFile("gen.bsh")
`

And what I can see by diff'ing the text is that the zsh and bash completion files for "GeneratedCli" contain all the artifacts for the full "GeneratedCli" command tree, while the completion files for "RootCmd" do not.  The two PS completion files are identical and only include artifacts for the "RootCmd".  So, based on experimental data, it seems like your comment about zsh/bash isn't right.

We're running v1.0.0 with your pull request applied.

marckhouzam commented:
Hi @rshoemaker.

With Cobra 1.0.0, completion for bash, zsh and powershell generate the exhaustive list of completions, as you have seen.  Completion for Fish shell however user a standard generic script that is the same for all commands and binaries.

For the next release (as currently on the master branch), zsh will start using this same approach of having a generic script, and the current powershell PR of @Luap99 also uses that standardized approach.  Bash completion will continue to work the old way, unless #1146 moves forward and also aligns bash completion with the other shells.

I haven't had time to look at your specific case but as pointed out by @Luap99, you may have a standardized way forward, which would be to start using `ValidArgsFunction` for your commands and have the commands call your sub-commands binaries with the `__complete` command.

We do this to handle `helm` plugins, where each plugin is a sub-command of the `helm` rootCmd, but delegates execution and completion to another binary.  Here is the code to delegate the completion to the plugin binary.  You may choose to adapte that code to call you sub-binaries with the `__complete` command:
https://github.com/helm/helm/blob/master/cmd/helm/load_plugins.go#L323-L377

As pointed out by @Luap99 you may need this PR to get flag-name completion to work: https://github.com/spf13/cobra/pull/1161

SilkyFowl commented on powershell_completions.go:
```diff
@@ -11,90 +8,278 @@ import (
+                        $Description = "  ($($comp.Description))"
+                    }
+
+                    [System.Management.Automation.CompletionResult]::new("$($comp.Name)$Description", "$($comp.Name)$Description", 'ParameterValue', "$($comp.Description)")
```
Why set `completionText` param to `"$($comp.Name)$Description"`? 
I couldn't figure out when I checked the code, but this seems to be the only constructor that adds a `$Description` to `completionText` parameter, is there any special meaning to it?

```pwsh
 $myParams = @{
  Native=$true
  CommandName='Foo'
}

$comps=@(
  @{n='add';d=':Add Foo'}
  @{n='list';d=':List Foo'}
  @{n='remove';d=':Remove Foo'}
  @{n='v';d=''}
)

Register-ArgumentCompleter @myParams -ScriptBlock {
  param($WordToComplete,$CommandAst,$CursorPosition)
    $comps.where{$_.n -match "^$WordToComplete"}  | % {
      $Mode = (Get-PSReadLineKeyHandler | Where-Object {$_.Key -eq "Tab" }).Function
        
      [CompletionResult]::new("$($_.n) $($_.d)", "$($_.n) $($_.d)", [CompletionResultType]::ParameterValue, " ")
    }
}
```

```console
~
❯ foo afdaf
add :Add Foo        list :List Foo      remove :Remove Foo  v

~
❯ foo add :Add Foo
```

Luap99 replied:
As you might know powershell support 3 different completion modes:
- `TabCompleteNext`
- `Complete`
- `MenuComplete`

By default (powershell) only `MenuComplete` will display a description below the suggestions.
Cobra allows descriptions in autocompletions for zsh and fish. PR #1146 adds description support to bash via a little hack. I adopted this hack here to also allow description to be displayed with the `Complete` mode since powershell does not support this natively . The idea is that the descriptions are appended to the real completion values. However if only one matching completion is left the script wont append the description so that the shell can autocomplete the correct value.

Users can choose if they want descriptions to be displayed while generating the scripts with `GenPowerShellCompletion` for no descriptions and  `GenPowerShellCompletionWithDesc` for descriptions.
```
# With descriptions and Mode 'Complete'
$ helm s[tab]
search  (search for a keyword in charts)  show  (show information of a chart)  status  (displays the status of the named release)
# if you type 'se' you will get this autocompleted to 'search'

# With descriptions and Mode 'MenuComplete' The description of the current selected value will be displayed below the suggestions.
$ helm s[tab]
search    show     status  
search for a keyword in charts

# Without descriptions
$ helm s[tab]
search  show  status
```

SilkyFowl replied:
Nice idea!!
I understand the intent of the code.
This is not a problem.

marckhouzam commented:
@jpmcb what do you think about merging this powershell completion patch?  From my testing @Luap99 did a spectacular job with this.  

Furthermore, it seems no one can get the existing powershell completion to work at all currently, so there is no risk of breaking anything 😄   In fact, I checked-out the commit from the original powershell completion pr #797 and still only got flag completion to work; it seems the feature never properly worked.

We need some good material for the next release and I thought this would be a great one.

jpmcb reviewed (approved):
Thanks for the time and patience on this one - I'll go ahead and merge since there seems to be a consensus that this is indeed working on windows / powershell. 

If anyone in the community is trying this out and finds issues, please feel free to open a new issue!!
